cgen error produced for if expression, that has a cast to an optional fn type

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.7 6eae77b.986e9cf
OS: linux, Ubuntu 20.04.6 LTS
Processor: 2 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz

getwd: /space/v/oo
vexe: /space/v/oo/v
vexe mtime: 2024-09-22 17:41:50

vroot: OK, value: /space/v/oo
VMODULES: OK, value: /home/delian/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.46.0
Git vroot status: weekly.2024.37-39-g986e9cfa (11 commit(s) behind V master)
.git/config present: true

CC version: cc (Ubuntu 10.5.0-1ubuntu1~20.04) 10.5.0
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9

```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg if_exprs_with_option_fn_type_cast.v`
```v
type DataFn = fn (name string) string
fn which_lang(name string) string { return name }
fn find_func(name string) ?DataFn {
	a := if name == 'vlang' { ?DataFn(which_lang) } else { none }
	return a
}

```

**What did you expect to see?**

a compiled program

**What did you see instead?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/if_exprs_with_option_fn_type_cast.01J8DDPYY9SJX7RM7XHWCYNC6V.tmp.c:13355: error: cannot convert 'struct _option_anon_fn_string__string' to 'struct string (*)(struct string)'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
```

Note: discovered while testing https://github.com/vlang/v/pull/22280 .
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -589,7 +589,7 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 				}
 				fn_name := c_fn_name(g.get_ternary_name(ident.name))
 
-				if val_type.has_flag(.option) && val is ast.SelectorExpr {
+				if val_type.has_flag(.option) && val in [ast.CastExpr, ast.SelectorExpr] {
 					ret_styp := g.typ(g.unwrap_generic(val_type))
 					g.write('${ret_styp} ${fn_name}')
 				} else {

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -589,7 +589,7 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 				}
 				fn_name := c_fn_name(g.get_ternary_name(ident.name))
 
-				if val_type.has_flag(.option) && val in [ast.CastExpr, ast.SelectorExpr] {
+				if val_type.has_flag(.option) {
 					ret_styp := g.typ(g.unwrap_generic(val_type))
 					g.write('${ret_styp} ${fn_name}')
 				} else {

--- a/vlib/v/tests/options/option_fntype_assign_test.v
+++ b/vlib/v/tests/options/option_fntype_assign_test.v
@@ -9,8 +9,16 @@ fn find_func(name string) ?DataFn {
 	return a
 }
 
+fn find_func2(name string) ?DataFn {
+	a := if name == 'vlang' { ?DataFn(which_lang) } else { none }
+	return a
+}
+
 fn test_main() {
 	if a := find_func('foo') {
 		assert a('bar') == 'bar'
+	}
+	if b := find_func('foo') {
+		assert b('bar') == 'bar'
 	}
 }

--- a/vlib/v/tests/options/option_fntype_assign_test.v
+++ b/vlib/v/tests/options/option_fntype_assign_test.v
@@ -1,0 +1,16 @@
+type DataFn = fn (name string) string
+
+fn which_lang(name string) string {
+	return name
+}
+
+fn find_func(name string) ?DataFn {
+	a := ?DataFn(which_lang)
+	return a
+}
+
+fn test_main() {
+	if a := find_func('foo') {
+		assert a('bar') == 'bar'
+	}
+}

